shared/common-adapters/tabs.tsx

```diff
@@ -1,70 +1,128 @@
 import * as React from 'react'
 import * as Styles from '../styles'
-import Box from './box'
+import Badge from './badge'
 import ClickableBox from './clickable-box'
 import Divider from './divider'
+import Icon, {IconType} from './icon'
+import ProgressIndicator from './progress-indicator'
+import Text from './text'
+import {Box, Box2} from './box'
+import capitalize from 'lodash/capitalize'
 
 const Kb = {
+  Badge,
   Box,
+  Box2,
   ClickableBox,
   Divider,
+  Icon,
+  ProgressIndicator,
+  Text,
 }
 
-type Props = {
-  clickableBoxStyle?: any
-  tabs: Array<React.ReactNode>
-  selected: React.ReactNode
-  onSelect: (s: React.ReactNode, idx: number) => any
-  style?: any
-  tabStyle?: any
+export type Tab<TitleT extends string> = {
+  title: TitleT
+  text?: string // text to show instead of title
+  icon?: IconType
+  badgeNumber?: number
 }
 
-const Tabs = ({clickableBoxStyle, tabs, selected, onSelect, style, tabStyle}: Props) => {
+type Props<TitleT extends string> = {
+  clickableBoxStyle?: Styles.StylesCrossPlatform
+  tabs: Array<Tab<TitleT>>
+  onSelect: (title: TitleT, idx: number) => void
+  selectedTab?: TitleT
+  style?: Styles.StylesCrossPlatform
+  tabStyle?: Styles.StylesCrossPlatform
+  showProgressIndicator?: boolean
+}
+
+const TabText = ({selected, text}: {selected: boolean; text: string}) => (
+  <Kb.Box2 style={styles.tabTextContainer} direction="horizontal">
+    <Kb.Text type="BodySmallSemibold" style={selected ? styles.selected : undefined}>
+      {text}
+    </Kb.Text>
+  </Kb.Box2>
+)
+
+function Tabs<TitleT extends string>(props: Props<TitleT>) {
   return (
-    <Kb.Box style={Styles.collapseStyles([styles.container, style])}>
-      {tabs.map((t, idx) => {
-        // @ts-ignore
-        const key: string = (t && t.key) || idx
+    <Kb.Box2
+      direction="horizontal"
+      style={Styles.collapseStyles([styles.container, props.style])}
+      alignItems="flex-start"
+      fullWidth={true}
+    >
+      {props.tabs.map((tab: Tab<TitleT>, idx: number) => {
+        const selected = props.selectedTab === tab.title
         return (
-          <Kb.ClickableBox onClick={() => onSelect(t, idx)} key={key} style={clickableBoxStyle}>
-            <Kb.Box style={styles.tabContainer}>
+          <Kb.ClickableBox
+            onClick={() => props.onSelect(tab.title, idx)}
```

Comment:
This can just be `onClick{props.onSelect}`

Reply:
no it can't, onSelect takes arguments.

Reply:
Doh, I'm an idiot. Carry on!